# Assignment 1 - All About torch.Tensor

### Deep Learning with PyTorch: Zero to Gans

An short introduction about PyTorch and about the chosen functions. 
- Fill Diagonal
- Eigenvalues
- Reshape
- Least Squares
- Unique

In [1]:
# Import torch and other required modules
import torch

## Fill Diagonal - fill_diagonal_()

According to the PyTorch Documentation, this function fills the main diagonal of a tensor (must be a matrix). Dimensions must be > 2

In [17]:
# Example 1 - working
test = torch.tensor([[1, 2, 3], 
                    [5, 43, 2],
                    [67, 5, 32]])

test.fill_diagonal_(10)

tensor([[10,  2,  3],
        [ 5, 10,  2],
        [67,  5, 10]])

Even in a nonzero matrix, the fill_diagonal_ function still diagnolizes the matrix to the specified parameter

In [15]:
# Example 2 - working
w = torch.zeros(6,6)
y = torch.zeros(7,7)
z = torch.zeros(8,8)
lst = [w, y, z]
a = 6

for i in lst:
    i.fill_diagonal_(a)
    a += 1
    print(i)

tensor([[6., 0., 0., 0., 0., 0.],
        [0., 6., 0., 0., 0., 0.],
        [0., 0., 6., 0., 0., 0.],
        [0., 0., 0., 6., 0., 0.],
        [0., 0., 0., 0., 6., 0.],
        [0., 0., 0., 0., 0., 6.]])
tensor([[7., 0., 0., 0., 0., 0., 0.],
        [0., 7., 0., 0., 0., 0., 0.],
        [0., 0., 7., 0., 0., 0., 0.],
        [0., 0., 0., 7., 0., 0., 0.],
        [0., 0., 0., 0., 7., 0., 0.],
        [0., 0., 0., 0., 0., 7., 0.],
        [0., 0., 0., 0., 0., 0., 7.]])
tensor([[8., 0., 0., 0., 0., 0., 0., 0.],
        [0., 8., 0., 0., 0., 0., 0., 0.],
        [0., 0., 8., 0., 0., 0., 0., 0.],
        [0., 0., 0., 8., 0., 0., 0., 0.],
        [0., 0., 0., 0., 8., 0., 0., 0.],
        [0., 0., 0., 0., 0., 8., 0., 0.],
        [0., 0., 0., 0., 0., 0., 8., 0.],
        [0., 0., 0., 0., 0., 0., 0., 8.]])


Diagnolizing zero matrices to create a stair like pattern

In [24]:
# Example 3 - breaking (to illustrate when it breaks)
test = torch.tensor(10)

test.fill_diagonal_(10.)

print(test)

RuntimeError: dimensions must larger than 1

As said before, the function requires that the dimensions of the matrix be more than 2. A single number tensor breaks the function.

This is a very useful function when you need to create certain matrices for machine learning.

## Eigenvalues - torch.eig()

An eigenvalue is the factor by which an eigenvector is stretched by a transformation within a system of equations.

torch.eig() finds the eigenvalues and eigenvectors of a real square matrix.

In [6]:
# Example 1 - working

x = torch.tensor([[8,3,10],
                 [9,4,1],
                 [5,2,50]], dtype=torch.float64)

torch.eig(x)

torch.return_types.eig(
eigenvalues=tensor([[ 0.4352,  0.0000],
        [10.2556,  0.0000],
        [51.3092,  0.0000]], dtype=torch.float64),
eigenvectors=tensor([], dtype=torch.float64))

As seen, torch.eig() found the eigen values of the square 3x3 matrix: .4352, 10.2556, 51.3092

In [3]:
# Example 2 - working

y = torch.randn(3,3)

torch.eig(y, eigenvectors=True)

torch.return_types.eig(
eigenvalues=tensor([[-0.0081,  0.9385],
        [-0.0081, -0.9385],
        [ 1.3846,  0.0000]]),
eigenvectors=tensor([[ 0.0777, -0.4747, -0.3504],
        [-0.8447,  0.0000,  0.5240],
        [-0.1027, -0.2110,  0.7763]]))

torch.eig() finds the eigenvectors by utilizing the eigenvectors parameter with a boolean value. If True, the eigenvectors are found as well, which can be seen in the above output.

In [11]:
# Example 3 - breaking (to illustrate when it breaks)

z = torch.tensor([[4],
                 [5],
                 [6]], dtype=torch.float64)

torch.eig(z)

RuntimeError: invalid argument 1: A should be square at /opt/conda/conda-bld/pytorch_1587428266983/work/aten/src/TH/generic/THTensorLapack.cpp:194

This vector breaks the function because eigenvalues can only be found in square matrices.

I've yet to learn the applications of eigenvalues for deep learning, but this must be an important concept in building these type of models.

## Reshape - torch.reshape()

Reshapes the dimensios of a tensor, so an n x m matrix can be converted into a m x n matrix.

In [13]:
# Example 1 - working

x = torch.randn(2,2)

print(torch.reshape(x, (1,4)))
torch.reshape(x, (4,1))

tensor([[0.1186, 1.6138, 2.7505, 0.1806]])


tensor([[0.1186],
        [1.6138],
        [2.7505],
        [0.1806]])

Reshapes the square matrix into a one-row matrix and into a vector.

In [16]:
# Example 2 - working

y = torch.randn(16,1)

torch.reshape(y, (4,4))

tensor([[ 0.1659,  0.1526,  0.4596, -0.2001],
        [ 0.0945, -0.0795, -0.0407, -0.8284],
        [ 0.3444, -0.9219,  0.9632, -0.1889],
        [-0.8507,  0.5135,  0.1139,  0.1020]])

Reshapes an entire vector into a neat 4x4 matrix. This emphasizes the need for the requested reshape to match the number of entities within the matrix.

In [17]:
# Example 3 - breaking (to illustrate when it breaks)

z = torch.randn(4,4)

torch.reshape(z, (5,5))

RuntimeError: shape '[5, 5]' is invalid for input of size 16

Example of what happens when the number of entities doesn't match the requested reshape.

Definitely a useful function for transforming a tensor to a shape to best fit certain matrix operations in machine learning.

## Least Squares

Finds solution to the least squares and least norm problems for a m x n matrix and a m x k matrix.

In [18]:
# Example 1 - working

x = torch.randn(5,5)
y = torch.randn(5,2)

torch.lstsq(y, x)

torch.return_types.lstsq(
solution=tensor([[ 1.7575, -0.2247],
        [ 1.7702, -0.6291],
        [-0.1232,  0.6687],
        [ 0.1176, -0.9535],
        [ 0.4702,  0.6696]]),
QR=tensor([[ 1.7401, -0.6656,  0.8801, -1.0346, -1.3289],
        [ 0.0209,  1.3381,  1.0780,  1.0486, -0.7973],
        [ 0.5526, -0.2711, -2.1116,  0.7179,  0.2614],
        [ 0.4871, -0.2520,  0.7736,  1.2741,  0.0740],
        [-0.2030,  0.2264, -0.5137, -0.0124,  0.6452]]))

If m >= n in x, then lstsq() returns the least squares solution 

In [20]:
# Example 2 - working

x = torch.randn(4,6)
y = torch.randn(4,3)

torch.lstsq(y,x)

torch.return_types.lstsq(
solution=tensor([[ 0.2770, -0.2452, -0.1787],
        [ 0.5487, -0.4738, -0.3252],
        [-0.6256,  0.1975, -0.0835],
        [ 0.2857, -0.4067, -0.5847],
        [-0.3278,  0.3651,  0.3471],
        [-0.3249, -0.1104, -0.4045]]),
QR=tensor([[-2.7082,  0.0495,  0.0041, -0.8697, -0.1465,  0.0413],
        [ 0.7410,  2.6154, -0.1434, -0.0650,  0.2551, -0.5636],
        [-0.7550, -1.0817,  1.5342, -0.3791,  0.4748, -0.2725],
        [-0.7039, -0.6512, -0.3462, -2.0219, -0.3021,  0.5164]]))

If m < n in x, then lstsq() returns least norm solution

In [21]:
# Example 3 - breaking (to illustrate when it breaks

x = torch.randn(3,3)
y = torch.randn(4,3)

torch.lstsq(y,x)

RuntimeError: Expected A and b to have same size at dim 0, but A has 3 rows and B has 4 rows

The lstsq() function breaks when the number of rows in the matrices do not match. This is because if the rows are the number of observations and the columns the number of features, than both matrices must have the same number of observations, otherwise the datasets must not be the same.

This is a useful function for understanding better the best fitting line of a regression model.

## Unique - torch.unique()

Returns the unique values in a tensor, which is presumably a vector or matrix

In [14]:
# Example 1 - working

x = torch.randn(3,3)

print(x)
torch.unique(x)

tensor([[ 0.4792,  0.4610,  0.3354],
        [-1.0780,  0.5112, -1.0263],
        [-0.2547, -0.4566, -1.0496]])


tensor([-1.0780, -1.0496, -1.0263, -0.4566, -0.2547,  0.3354,  0.4610,  0.4792,
         0.5112])

A simple example of the funtion in action, returning the unique values of x.

In [4]:
# Example 2 - working

w = torch.tensor([[2,2,3],
             [2,4,2],
             [1,2,2]])

y = torch.tensor([[3,3,2],
                  [3,1,3],
                  [4,3,3]])

print(torch.unique(w))
print(torch.unique(y))

solution = 1**2 + 2**2 + 3**2 + 4**2

product = torch.unique(w) * torch.unique(y)

print(product)
print('Solution: ', solution)

torch.sum(product) == solution

tensor([1, 2, 3, 4])
tensor([1, 2, 3, 4])
tensor([ 1,  4,  9, 16])
Solution:  30


tensor(True)

Here I created tensors that have their 2 and 3 as modes and a unique value in each row. I wanted to play around with mathematical and conditional operators with PyTorch so I wrote the above code to see if the sum of the unique values of w and y would be equal to the solution, 30. Seeing this in action was really interesting.

In [15]:
# Example 3 - breaking (to illustrate when it breaks)

z = torch.zeros(5,5)

torch.unique(z)

test = torch.tensor([[5,4,3],
                     [6,7,8],
                     [9,10,11]])
del(test)

print(z)
torch.unique(test)

tensor([[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]])


NameError: name 'test' is not defined

The unique function works even with a zero matrix. To break it, the variable the function is called for must not exist/ be undefined. We can see this when I delete the test variable.

Finding unique values in tensors is important, I imagine for the same reasons it is important for finding them in a dataset: to avoid duplicates. Duplicates can skew the results of a deep learning model and prior analyses.

## Conclusion

After covering both common and linear algebra functions in PyTorch, I can see I still have a lot to learn and discover about how all of these functions collaborate to build neural networks and accomplish machine learning tasks. I definitely plan to revist the PyTorch documentation for more interesting functions to play with. 

## Reference Links
Provide links to your references and other interesting articles about tensors
* Official documentation for `torch.Tensor`: https://pytorch.org/docs/stable/tensors.html
* ...

In [27]:
!pip install jovian --upgrade --quiet

In [28]:
import jovian

In [ ]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Please enter your API key ( from https://jovian.ml/ ):
API KEY: 